Implementing #99

HaoTNN opened the pull request:
Change some things around to make it possible as well. I kind of want to move the logic out of Core_Client, but up to you. Lemme know what you think.

HaoTNN committed:
Implementing #99

HaoTNN committed:
Changed default of CommandsRequireClassifier to true

zrisher commented on Data/Scripts/GardenConquest/Extensions/GridExtensions.cs:
```diff
@@ -184,6 +186,48 @@ public static class GridExtensions {
+
+			// Get all beacon blocks
+			Func<IMySlimBlock, bool> isBeacon = b => b.FatBlock != null && b.FatBlock is InGame.IMyBeacon;
+			grid.GetBlocks(beaconBlocks, isBeacon);
```
Would you mind defining isBeacon inline within the call to grid.GetBlocks? Either that, or assign it outside of this function, so that we don't have to do new object allocation every time for this. I.e.

```
grid.GetBlocks(beaconBlocks, (b => b.FatBlock != null && b.FatBlock is InGame.IMyBeacon));
```

zrisher commented on Data/Scripts/GardenConquest/Extensions/GridExtensions.cs:
```diff
@@ -184,6 +186,48 @@ public static class GridExtensions {
+				String subTypeName = block.FatBlock.BlockDefinition.SubtypeName;
+				if (subTypeName.Contains(Blocks.HullClassifier.SHARED_SUBTYPE)) {
+					return block.FatBlock;
+				}
```
I think the stuff in this loop could make a nice extension function on IMySlimBlock, i.e.

```
foreach (IMySlimBlock block in beaconBlocks) {
    if (block.IsClassifier()) {
        return block.FatBlock;
    }
}
```

HaoTNN replied:
There's `isClassifierBlock(IMySlimBlock block)` in HullClassifier, should we just use that or  move it to a new class `IMySlimBlockExtensions.cs`?

zrisher replied:
Yep sounds perfect.

zrisher commented on Data/Scripts/GardenConquest/Core/ConquestSettings.cs:
```diff
@@ -28,6 +28,7 @@ public struct SETTINGS {
 			public List<ControlPoint> ControlPoints { get; set; }
 			public int CPPeriod { get; set; }
 			public int CleanupPeriod { get; set; }
+			public bool CommandsRequireClassifier { get; set; }
```
Would you please rename this to `SimpleOwnership`? We need it to indicate that we use classifier for ownership checks not just for commands but for `GridEnforcer.reevaluateOwnership()` as well. We want to do the latter to take advantage of the fact that thieves have to grind classifiers now, so we can do the ownership flip immediately there, instead of people saying "Someone stole my ship but it's still counting against my limits!".

zrisher commented on Data/Scripts/GardenConquest/Messaging/RequestProcessor.cs:
```diff
@@ -171,10 +171,12 @@ public class RequestProcessor {
 					case GridOwner.OWNER_TYPE.FACTION:
-						errorMessage = "Your faction does not have control of that ship's Main Cockpit!";
+						errorMessage = "Your faction does not have control of that ship's ";
+						errorMessage += (ConquestSettings.getInstance().CommandsRequireClassifier == true ? "Hull Classifier!" : "Main Cockpit!");
```
This should be unnecessary in the case of simple ownership, because nothing will be on the the fleet list that they don't already have control of.

zrisher commented on Data/Scripts/GardenConquest/Messaging/ResponseProcessor.cs:
```diff
@@ -313,7 +313,8 @@ public class ResponseProcessor {
 							else {
-								fleetInfoBody += "      GPS: Unavailable - Must own the Main Cockpit\n";
+								fleetInfoBody += "      GPS: Unavailable - Must own the ";
+								fleetInfoBody += (m_ServerSettings.CommandsRequireClassifier == true ? "Hull Classifier\n" : "Main Cockpit\n");
```
Same as above, the check itself is unnecessary in the case of simple ownership.

zrisher commented:
Per the above comments, would you please also add a check to `GridEnforcer.reevaluateOwnership()` for `SimpleOwnership` and just pass the classifier owner instead of of those of all blocks?

zrisher commented:
Let's wait on moving things out of core client until I refactor it to use the new session components from SEGarden.

HaoTNN commented:
Regarding the check to GridEnforcer.reevaluateOwnership():

I.e,

```
if (s_Settings.SimpleOwnership) {
    changed = m_Owner.reevaluateOwnership(new List<long> { m.Grid.getClassifierBlock().OwnerId });
}
else {
    changed = m_Owner.reevaluateOwnership(BigOwners);
}
```

Right?

HaoTNN committed:
Implemented suggested changes

HaoTNN committed:
Further changes

zrisher commented:
This has been merged and fixed up, but not released to public yet because I'm still getting some residual errors. Currently, if I create a grid as a player, switch to a faction, add/remove some blocks, and try to change classifier to one that requires a faction I get a "you're not in a faction" message.

zrisher commented:
Fixed with #130, will test and then release.
